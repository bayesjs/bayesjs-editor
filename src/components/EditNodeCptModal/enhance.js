import { connectify } from 'decorators';
import {
  branch,
  compose,
  renderNothing,
  withProps,
  withHandlers,
} from 'recompose';
import {
  isNil,
  prop,
  pipe,
  isEmpty,
  complement,
} from 'ramda';
import { getEditingNodeCpt } from 'selectors/editing-node-cpt';
import { onCancelEditingNodeCpt, onSaveEditingNodeCpt } from 'actions/editing-node-cpt';

const hasNotNode = pipe(prop('node'), isNil);
const isNotEmpty = complement(isEmpty);

const enhance = compose(
  connectify({
    node: getEditingNodeCpt,
  }, {
    onSave: onSaveEditingNodeCpt,
    onCancel: onCancelEditingNodeCpt,
  }),
  branch(hasNotNode, renderNothing),
  withProps(({ node: { parents } }) => ({
    hasParents: isNotEmpty(parents),
  })),
  withHandlers({
    onAlert: () => window.alert,
  }),
);

export default enhance;
