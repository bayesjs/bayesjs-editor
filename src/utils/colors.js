import { sort, isEmpty } from 'ramda';

export const COLORS = [
  '#D84315',
  '#BF360C',
  '#FBC02D',
  '#F9A825',
  '#F57F17',
  '#FFA000',
  '#FF8F00',
  '#FF6F00',
  '#F57C00',
  '#EF6C00',
  '#E65100',
  '#388E3C',
  '#2E7D32',
  '#689F38',
  '#558B2F',
  '#AFB42B',
  '#0288D1',
  '#0097A7',
  '#00838F',
  '#1976D2',
  '#1565C0',
  '#C2185B',
];

const shuffle = sort(() => Math.random() - 0.5);

const createRandomColorBuilder = () => {
  let list = [];
  return () => {
    if (isEmpty(list)) {
      list = shuffle(COLORS);
    }
    return list.shift();
  };
};

export const getRandomColor = createRandomColorBuilder();
