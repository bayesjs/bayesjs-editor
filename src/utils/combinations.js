import {
  assoc,
  forEach,
  head,
  isEmpty,
  tail,
} from 'ramda';

const createCombinations = (onCreateCombination, allNodes, acc = {}) => {
  if (isEmpty(allNodes)) {
    onCreateCombination(acc);
  } else {
    const { id, states } = head(allNodes);

    forEach((state) => {
      createCombinations(onCreateCombination, tail(allNodes), assoc(id, state, acc));
    }, states);
  }
};

export const createNodeCombinations = (nodes) => {
  const combinations = [];

  createCombinations(
    combination => combinations.push(combination),
    nodes,
  );

  return combinations;
};
