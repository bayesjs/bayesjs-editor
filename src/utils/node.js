import {
  find,
  findIndex,
  identity,
  propEq,
  reject,
  useWith,
} from 'ramda';

const propId = propEq('id');

export const findNodeById = useWith(
  find,
  [propId, identity],
);

export const findIndexNodeById = useWith(
  findIndex,
  [propId, identity],
);

export const removeNodeById = useWith(
  reject,
  [propId, identity],
);
