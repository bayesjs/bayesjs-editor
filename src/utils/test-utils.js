import {
  compose,
  filter,
  isEmpty,
  join,
  map,
  not,
} from 'ramda';

import { kebabCase } from 'change-case';
import removeAccents from 'remove-accents';

const formatWord = compose(removeAccents, kebabCase);
const notEmpty = compose(not, isEmpty);
const filterSubItems = filter(notEmpty);
const formatSubItems = compose(join('-'), map(formatWord));

export const getComponentTestId = (componentName, ...subItems) => { //eslint-disable-line
  const compName = formatWord(componentName);
  const filtredItems = filterSubItems(subItems);

  return notEmpty(filtredItems) ? `${compName}__${formatSubItems(filtredItems)}` : compName;
};
