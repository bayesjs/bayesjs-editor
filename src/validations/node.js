import {
  all,
  allPass,
  any,
  compose,
  equals,
  has,
  ifElse,
  is,
  isEmpty,
  not,
  pipe,
  prop,
  sum,
  useWith,
  values,
} from 'ramda';

import { roundValue } from '@utils/math';

const isArray = is(Array);
const isObject = is(Object);
const sumValues = compose(sum, values);
const equalsOne = equals(1);
const propThen = prop('then');

const propIsArray = myProp => pipe(prop(myProp), isArray);
const propIsNotArray = myProp => pipe(prop(myProp), isArray, not);
const propIsObject = myProp => pipe(prop(myProp), isObject);

export const hasConnections = allPass([
  has('linkedNode'),
  propIsNotArray('linkedNode'),
  propIsObject('linkedNode'),
]);

export const hasStates = allPass([has('states'), propIsArray('states')]);

export const hasDescription = ({ showDescription, description }) =>
  Boolean(showDescription && description);

const isSumValuesEqualsOne = pipe(sumValues, roundValue, equalsOne);
const isAllThenSumValuesEqualsOne = all(pipe(propThen, isSumValuesEqualsOne));

export const isNodeCptValid = ifElse(
  isArray,
  isAllThenSumValuesEqualsOne,
  isSumValuesEqualsOne,
);

export const isNodeWithoutParents = pipe(prop('parents'), isEmpty);

export const containsParentInNode = useWith(any, [equals, prop('parents')]);
