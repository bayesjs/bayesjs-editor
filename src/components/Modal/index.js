import PropTypes from 'prop-types';
import React from 'react';
import ReactModal from 'react-modal';
import classNames from 'classnames';
import fontAwesome from 'font-awesome/css/font-awesome.css';
import { getComponentTestId } from '../../utils/test-utils';
import styles from './styles.css';

const style = {
  overlay: {
    backgroundColor: 'rgba(50, 50, 50, 0.75)',
  },
  content: {
    top: '50%',
    left: '50%',
    right: 'auto',
    bottom: 'auto',
    transform: 'translate(-50%, -50%)',
    border: 'solid 1px #333',
    borderRadius: 0,
    padding: 0,
    margin: 0,
  },
};

const Modal = ({
  title,
  isOpen,
  onRequestClose,
  children,
  name,
}) => (
  <ReactModal
    isOpen={isOpen}
    onRequestClose={onRequestClose}
    style={style}
    contentLabel={title}
  >
    <div className={styles.header} data-testid={getComponentTestId('ModalHeader', name || title)}>
      {title}
      <i
        onClick={onRequestClose}
        className={classNames([fontAwesome.fa, fontAwesome.faTimes, styles.closeButton])}
        title="Fechar"
      />
    </div>

    <div className={styles.body} data-testid={getComponentTestId('ModalBody', name || title)}>
      {children}
    </div>
  </ReactModal>
);

Modal.defaultProps = {
  children: null,
  name: '',
};

Modal.propTypes = {
  title: PropTypes.string.isRequired,
  isOpen: PropTypes.bool.isRequired,
  onRequestClose: PropTypes.func.isRequired,
  children: PropTypes.element,
  name: PropTypes.string,
};

export default Modal;
