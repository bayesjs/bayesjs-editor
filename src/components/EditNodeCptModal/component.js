import React, { useCallback, useState } from 'react';

import Button from '@components/Button';
import Modal from '@components/Modal';
import NodeCptEditTable from '@components/NodeCptEditTable';
import NodeCptParentStatesTable from '@components/NodeCptParentStatesTable';
import PropTypes from 'prop-types';
import { isEnterKey } from '@utils/event';
import { isNodeCptValid } from '@validations/node';
import { nodePropTypes } from '@models';
import styles from './styles.scss';

const handleCptKeyUpCreator = handleSave => (e) => {
  if (isEnterKey(e)) {
    handleSave();
  }
};

const handleSaveCreator = (id, cpt, onSave, onAlert) => () => {
  if (isNodeCptValid(cpt)) {
    onSave(id, cpt);
  } else {
    onAlert('A soma das probabilidades para cada uma das linhas deve ser igual a 1');
  }
};

const EditNodeCptModal = ({
  node,
  hasParents,
  onSave,
  onCancel,
  onAlert,
}) => {
  const { id, states } = node;
  const [cpt, setCpt] = useState(node.cpt);
  const handleSave = useCallback(handleSaveCreator(id, cpt, onSave, onAlert), [id, cpt]);
  const handleCptKeyUp = handleCptKeyUpCreator(handleSave);

  return (
    <Modal
      title={`Editar Tabela de Probabilidades (${id})`}
      onRequestClose={onCancel}
      isOpen
    >
      <div className={styles.tablesContainer}>
        {hasParents && <NodeCptParentStatesTable node={node} />}
        <NodeCptEditTable
          cpt={cpt}
          hasParents={hasParents}
          states={states}
          onChange={setCpt}
          onKeyUp={handleCptKeyUp}
        />
      </div>

      <div className={styles.buttonsContainer}>
        <Button onClick={handleSave} primary>
          Salvar
        </Button>

        <Button onClick={onCancel}>
          Cancelar
        </Button>
      </div>
    </Modal>
  );
};

EditNodeCptModal.propTypes = {
  node: nodePropTypes,
  onSave: PropTypes.func.isRequired,
  onCancel: PropTypes.func.isRequired,
  onAlert: PropTypes.func.isRequired,
  hasParents: PropTypes.bool.isRequired,
};

EditNodeCptModal.defaultProps = {
  node: null,
};

export default EditNodeCptModal;
