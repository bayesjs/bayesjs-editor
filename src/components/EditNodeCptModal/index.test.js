import React from 'react';
import { render, fireEvent } from '@testing-library/react';
import { getComponentTestId } from 'utils/test-utils';
import { head } from 'ramda';
import Modal from 'react-modal';
import * as nodeValidation from 'validations/node';
import * as eventUtils from 'utils/event';
import EditNodeCptModal from './component';

const renderComponent = (props) => {
  Modal.setAppElement(document.body);

  return {
    ...render(<EditNodeCptModal {...props} />),
    element: document.body.lastChild,
  };
};

describe('EditNodeCptModal Component', () => {
  let component;

  describe('When node has no parents', () => {
    const props = {
      node: {
        id: 'Node 1',
        states: [
          'True',
          'False',
        ],
        parents: [],
        cpt: {
          True: 0.5,
          False: 0.5,
        },
      },
      hasParents: false,
      onSave: jest.fn(),
      onCancel: jest.fn(),
      onAlert: jest.fn(),
    };

    beforeEach(() => {
      component = renderComponent(props);
    });

    it('matches snapshot', () => {
      expect(component.element).toMatchSnapshot();
    });

    describe('When key pressed in cpt input', () => {
      beforeEach(() => {
        props.onSave.mockClear();
      });

      describe('and is "enter" key', () => {
        const { isEnterKey } = eventUtils;

        beforeEach(() => {
          const { getAllByTestId } = component;
          const inputCptTrueState = head(getAllByTestId(getComponentTestId('InputCpt', 'True')));

          eventUtils.isEnterKey = () => true;
          fireEvent.keyUp(inputCptTrueState);
        });

        afterAll(() => {
          eventUtils.isEnterKey = isEnterKey;
        });

        it('calls onSave fucntion', () => {
          expect(props.onSave).toHaveBeenCalledWith(props.node.id, props.node.cpt);
        });
      });

      describe('and is not "enter" key', () => {
        const { isEnterKey } = eventUtils;

        beforeEach(() => {
          const { getAllByTestId } = component;
          const inputCptTrueState = head(getAllByTestId(getComponentTestId('InputCpt', 'True')));

          eventUtils.isEnterKey = () => false;
          fireEvent.keyUp(inputCptTrueState);
        });

        afterAll(() => {
          eventUtils.isEnterKey = isEnterKey;
        });

        it('does not call onSave fucntion', () => {
          expect(props.onSave).not.toHaveBeenCalled();
        });
      });
    });

    describe('When saving', () => {
      beforeEach(() => {
        props.onSave.mockClear();
        props.onAlert.mockClear();
      });

      describe('and cpt is valid', () => {
        beforeEach(() => {
          const { getByText } = component;
          const saveButton = getByText('Salvar');

          fireEvent.click(saveButton);
        });

        it('calls onSave function with cpt', () => {
          expect(props.onSave).toHaveBeenCalledWith(props.node.id, props.node.cpt);
        });
      });

      describe('and cpt is invalid', () => {
        const { isNodeCptValid } = nodeValidation;

        beforeEach(() => {
          const { getByText } = component;
          const saveButton = getByText('Salvar');

          nodeValidation.isNodeCptValid = () => false;
          fireEvent.click(saveButton);
        });

        afterAll(() => {
          nodeValidation.isNodeCptValid = isNodeCptValid;
        });

        it('calls onAlert function with message', () => {
          expect(props.onAlert).toHaveBeenCalledWith('A soma das probabilidades para cada uma das linhas deve ser igual a 1');
        });
      });
    });
  });


  describe('When node has parents', () => {
    const props = {
      node: {
        id: 'Node 3',
        states: [
          'True',
          'False',
        ],
        parents: [
          'Node 2',
          'Node 1',
        ],
        cpt: [
          {
            when: {
              'Node 2': 'True',
              'Node 1': 'True',
            },
            then: {
              True: 0.5,
              False: 0.5,
            },
          },
          {
            when: {
              'Node 2': 'False',
              'Node 1': 'True',
            },
            then: {
              True: 0.5,
              False: 0.5,
            },
          },
          {
            when: {
              'Node 2': 'True',
              'Node 1': 'False',
            },
            then: {
              True: 0.5,
              False: 0.5,
            },
          },
          {
            when: {
              'Node 2': 'False',
              'Node 1': 'False',
            },
            then: {
              True: 0.5,
              False: 0.5,
            },
          },
        ],
      },
      hasParents: true,
      onSave: jest.fn(),
      onCancel: jest.fn(),
      onAlert: jest.fn(),
    };

    beforeEach(() => {
      component = renderComponent(props);
    });

    it('matches snapshot', () => {
      expect(component.element).toMatchSnapshot();
    });

    describe('When key pressed in cpt input', () => {
      beforeEach(() => {
        props.onSave.mockClear();
      });

      describe('and is "enter" key', () => {
        const { isEnterKey } = eventUtils;

        beforeEach(() => {
          const { getAllByTestId } = component;
          const inputCptTrueState = head(getAllByTestId(getComponentTestId('InputCpt', 'True')));

          eventUtils.isEnterKey = () => true;
          fireEvent.keyUp(inputCptTrueState);
        });

        afterAll(() => {
          eventUtils.isEnterKey = isEnterKey;
        });

        it('calls onSave fucntion', () => {
          expect(props.onSave).toHaveBeenCalledWith(props.node.id, props.node.cpt);
        });
      });

      describe('and is not "enter" key', () => {
        const { isEnterKey } = eventUtils;

        beforeEach(() => {
          const { getAllByTestId } = component;
          const inputCptTrueState = head(getAllByTestId(getComponentTestId('InputCpt', 'True')));

          eventUtils.isEnterKey = () => false;
          fireEvent.keyUp(inputCptTrueState);
        });

        afterAll(() => {
          eventUtils.isEnterKey = isEnterKey;
        });

        it('does not call onSave fucntion', () => {
          expect(props.onSave).not.toHaveBeenCalled();
        });
      });
    });

    describe('When saving', () => {
      beforeEach(() => {
        props.onSave.mockClear();
        props.onAlert.mockClear();
      });

      describe('and cpt is valid', () => {
        beforeEach(() => {
          const { getByText } = component;
          const saveButton = getByText('Salvar');

          fireEvent.click(saveButton);
        });

        it('calls onSave function with cpt', () => {
          expect(props.onSave).toHaveBeenCalledWith(props.node.id, props.node.cpt);
        });
      });

      describe('and cpt is invalid', () => {
        const { isNodeCptValid } = nodeValidation;

        beforeEach(() => {
          const { getByText } = component;
          const saveButton = getByText('Salvar');

          nodeValidation.isNodeCptValid = () => false;
          fireEvent.click(saveButton);
        });

        afterAll(() => {
          nodeValidation.isNodeCptValid = isNodeCptValid;
        });

        it('calls onAlert function with message', () => {
          expect(props.onAlert).toHaveBeenCalledWith('A soma das probabilidades para cada uma das linhas deve ser igual a 1');
        });
      });
    });
  });
});
