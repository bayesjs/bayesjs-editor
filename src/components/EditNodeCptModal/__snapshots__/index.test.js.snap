// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EditNodeCptModal Component When node has no parents matches snapshot 1`] = `
<div
  class="ReactModalPortal"
>
  <div
    class="ReactModal__Overlay ReactModal__Overlay--after-open"
    style="position: fixed; top: 0px; left: 0px; right: 0px; bottom: 0px; background-color: rgba(50, 50, 50, 0.75);"
  >
    <div
      aria-label="Editar Tabela de Probabilidades (Node 1)"
      class="ReactModal__Content ReactModal__Content--after-open"
      role="dialog"
      style="position: absolute; top: 50%; left: 50%; border: 1px solid #333; background: rgb(255, 255, 255); overflow: auto; border-radius: 0; outline: none; padding: 0px; transform: translate(-50%, -50%); margin: 0px;"
      tabindex="-1"
    >
      <div
        class="header"
        data-testid="modal-header__editar-tabela-de-probabilidades-node-1"
      >
        Editar Tabela de Probabilidades (Node 1)
        <i
          class="fa faTimes closeButton"
          title="Fechar"
        />
      </div>
      <div
        class="body"
        data-testid="modal-body__editar-tabela-de-probabilidades-node-1"
        style="max-width: 819.2px; max-height: 614.4000000000001px;"
      >
        <div
          class="tablesContainer"
        >
          <table>
            <thead>
              <tr>
                <th>
                  True
                </th>
                <th>
                  False
                </th>
              </tr>
            </thead>
            <tbody>
              <tr
                class=""
              >
                <td>
                  <input
                    class="inputCpt"
                    data-testid="input-cpt__true"
                    max="1"
                    min="0"
                    step="0.01"
                    type="number"
                    value="0.5"
                  />
                </td>
                <td>
                  <input
                    class="inputCpt"
                    data-testid="input-cpt__false"
                    max="1"
                    min="0"
                    step="0.01"
                    type="number"
                    value="0.5"
                  />
                </td>
              </tr>
            </tbody>
          </table>
        </div>
        <div
          class="buttonsContainer"
        >
          <button
            class="--primary"
            data-testid="button"
            title=""
            type="button"
          >
            Salvar
          </button>
          <button
            class=""
            data-testid="button"
            title=""
            type="button"
          >
            Cancelar
          </button>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`EditNodeCptModal Component When node has parents matches snapshot 1`] = `
<div
  class="ReactModalPortal"
>
  <div
    class="ReactModal__Overlay ReactModal__Overlay--after-open"
    style="position: fixed; top: 0px; left: 0px; right: 0px; bottom: 0px; background-color: rgba(50, 50, 50, 0.75);"
  >
    <div
      aria-label="Editar Tabela de Probabilidades (Node 3)"
      class="ReactModal__Content ReactModal__Content--after-open"
      role="dialog"
      style="position: absolute; top: 50%; left: 50%; border: 1px solid #333; background: rgb(255, 255, 255); overflow: auto; border-radius: 0; outline: none; padding: 0px; transform: translate(-50%, -50%); margin: 0px;"
      tabindex="-1"
    >
      <div
        class="header"
        data-testid="modal-header__editar-tabela-de-probabilidades-node-3"
      >
        Editar Tabela de Probabilidades (Node 3)
        <i
          class="fa faTimes closeButton"
          title="Fechar"
        />
      </div>
      <div
        class="body"
        data-testid="modal-body__editar-tabela-de-probabilidades-node-3"
        style="max-width: 819.2px; max-height: 614.4000000000001px;"
      >
        <div
          class="tablesContainer"
        >
          <table>
            <thead>
              <tr>
                <th>
                  Node 2
                </th>
                <th>
                  Node 1
                </th>
              </tr>
            </thead>
            <tbody>
              <tr>
                <td>
                  True
                </td>
                <td>
                  True
                </td>
              </tr>
              <tr>
                <td>
                  False
                </td>
                <td>
                  True
                </td>
              </tr>
              <tr>
                <td>
                  True
                </td>
                <td>
                  False
                </td>
              </tr>
              <tr>
                <td>
                  False
                </td>
                <td>
                  False
                </td>
              </tr>
            </tbody>
          </table>
          <table>
            <thead>
              <tr>
                <th>
                  True
                </th>
                <th>
                  False
                </th>
              </tr>
            </thead>
            <tbody>
              <tr
                class=""
              >
                <td>
                  <input
                    class="inputCpt"
                    data-testid="input-cpt__true"
                    max="1"
                    min="0"
                    step="0.01"
                    type="number"
                    value="0.5"
                  />
                </td>
                <td>
                  <input
                    class="inputCpt"
                    data-testid="input-cpt__false"
                    max="1"
                    min="0"
                    step="0.01"
                    type="number"
                    value="0.5"
                  />
                </td>
              </tr>
              <tr
                class=""
              >
                <td>
                  <input
                    class="inputCpt"
                    data-testid="input-cpt__true"
                    max="1"
                    min="0"
                    step="0.01"
                    type="number"
                    value="0.5"
                  />
                </td>
                <td>
                  <input
                    class="inputCpt"
                    data-testid="input-cpt__false"
                    max="1"
                    min="0"
                    step="0.01"
                    type="number"
                    value="0.5"
                  />
                </td>
              </tr>
              <tr
                class=""
              >
                <td>
                  <input
                    class="inputCpt"
                    data-testid="input-cpt__true"
                    max="1"
                    min="0"
                    step="0.01"
                    type="number"
                    value="0.5"
                  />
                </td>
                <td>
                  <input
                    class="inputCpt"
                    data-testid="input-cpt__false"
                    max="1"
                    min="0"
                    step="0.01"
                    type="number"
                    value="0.5"
                  />
                </td>
              </tr>
              <tr
                class=""
              >
                <td>
                  <input
                    class="inputCpt"
                    data-testid="input-cpt__true"
                    max="1"
                    min="0"
                    step="0.01"
                    type="number"
                    value="0.5"
                  />
                </td>
                <td>
                  <input
                    class="inputCpt"
                    data-testid="input-cpt__false"
                    max="1"
                    min="0"
                    step="0.01"
                    type="number"
                    value="0.5"
                  />
                </td>
              </tr>
            </tbody>
          </table>
        </div>
        <div
          class="buttonsContainer"
        >
          <button
            class="--primary"
            data-testid="button"
            title=""
            type="button"
          >
            Salvar
          </button>
          <button
            class=""
            data-testid="button"
            title=""
            type="button"
          >
            Cancelar
          </button>
        </div>
      </div>
    </div>
  </div>
</div>
`;
